cimgen/languages/modernpython/lang_pack.py

```diff
@@ -76,23 +80,124 @@ def _get_type_and_default(text, render) -> tuple[str, str]:
         return ("str", 'default=""')
 
 
+def _primitive_to_data_type(datatype):
+    if datatype.lower() == "integer":
+        return "int"
+    if datatype.lower() == "float":
+        return "float"
+    if datatype.lower() == "boolean":
+        return "bool"
+    if datatype.lower() == "datetime":
+        return "datetime"
+    if datatype.lower() == "date":
+        return "date"
+    if datatype.lower() == "time":
+        return "time"
+    if datatype.lower() == "monthday":
+        return "str"  # TO BE FIXED? I could not find a datatype in python that holds only month and day.
+    if datatype.lower() == "string":
+        return "str"
+    else:
+        # this actually never happens
+        return "float"
+
+
+def _set_imports(attributes):
+    classes = set()
+    for attribute in attributes:
+        if attribute["is_datatype_attribute"] or attribute["is_primitive_attribute"]:
+            classes.add(attribute["attribute_class"])
+
+    result = ""
+    for val in sorted(classes):
+        result += "from ." + val + " import " + val + "\n"
+    return result + "\n"
+
+
+def _set_datatype_attributes(attributes) -> dict:
+    datatype_attributes = {}
+    datatype_attributes["python_type"] = "None"
+    datatype_attributes["unit"] = "UnitSymbol.none"
+    datatype_attributes["multiplier"] = "UnitMultiplier.none"
```

Comment:
hi @tom-hg57 
i think since at least version 3.0 (and probably also newer version of 2.x) multipliers are fixed, which means that indeed, you won't find a profile instance specifying the unit.

about adding denominatorUnit and denominatorMultiplier, we didn't look into that since we work mostly on CGMES 3, but if you think it's good to have them, it should not be too much effort to add a flag for them in case of CGMES 2.0.

Reply:
I didn't use CapacitancePerLength so far. So it's surely not so importent.

Otherwise, maybe it could make _set_datatype_attributes even simpler to go over all attributes and handle all with "isFixed". What do you think?

Reply:
to provide a constructive feedback, i need to look into the code, allow me some time, I am on another project today :)

Reply:
I found a solution:
```
def _set_datatype_attributes(attributes) -> dict:
    datatype_attributes = {}
    datatype_attributes["python_type"] = "None"
    datatype_attributes["unit"] = "UnitSymbol.none"
    datatype_attributes["multiplier"] = "UnitMultiplier.none"

    for attribute in attributes:
        if attribute["is_datatype_attribute"] or attribute["is_primitive_attribute"]:
            datatype_attributes["python_type"] = _primitive_to_data_type(attribute["attribute_class"])
        if "isFixed" in attribute:
            datatype_attributes[attribute["label"]] = attribute["attribute_class"] + "." + attribute["isFixed"]
    return datatype_attributes
```
```
{{class_name}} = CIMDatatype(
    name="{{class_name}}",
    type={{python_type}},
{{#unit}}
    symbol={{unit}},
{{/unit}}
{{#multiplier}}
    multiplier={{multiplier}},
{{/multiplier}}
{{#denominatorUnit}}
    denominatorUnit={{denominatorUnit}},
{{/denominatorUnit}}
{{#denominatorMultiplier}}
    denominatorMultiplier={{denominatorMultiplier}},
{{/denominatorMultiplier}}
    profiles=[{{#class_origin}}
        Profile.{{origin}},{{/class_origin}}
    ],
)
```
The results for CGMES3 would be the same.

But there is a problem in Money.py:
```
Money = CIMDatatype(
    name="Money",
    type=float,
    symbol=UnitSymbol.none,
    multiplier=UnitMultiplier.none,
    profiles=[
        Profile.EQ,
    ],
)
```
1. the attribute type for "symbol" (= "unit") is not UnitSymbol, but Currency
2. attribute "unit" is not fixed for Money

The better (and simpler) solution is:
```
{{class_name}} = CIMDatatype(
    name="{{class_name}}",
    type={{python_type}},
{{#attributes}}
{{#isFixed}}
    {{label}}={{attribute_class}}.{{isFixed}},
{{/isFixed}}
{{/attributes}}
    profiles=[{{#class_origin}}
        Profile.{{origin}},{{/class_origin}}
    ],
)
```
```
def _set_datatype_attributes(attributes) -> dict:
    datatype_attributes = {}
    datatype_attributes["python_type"] = "None"

    for attribute in attributes:
        if attribute["is_datatype_attribute"] or attribute["is_primitive_attribute"]:
            datatype_attributes["python_type"] = _primitive_to_data_type(attribute["attribute_class"])
    return datatype_attributes
```
Result for CGMES3:
```
Money = CIMDatatype(
    name="Money",
    type=float,
    multiplier=UnitMultiplier.none,
    profiles=[
        Profile.EQ,
    ],
)
```
without symbol/unit. And there is now "unit" instead of "symbol" for all other datatype classes.
If this is not a problem for your reader/writer you could use the second (simpler) solution.

Reply:
Your second solution is fine by me.

Some changes are however required to the class CIMDatatype.
- Addition of attributes: `denominatorMultiplier` and `denominatorUnit`
- In some of the cgmes schema (2.4) the `Money.unit` can be fixed, therefore the `Currency` class should be added as acceptable type for `unit`

```
@dataclass(config=cgmes_resource_config)
class CIMDatatype(Primitive):

    multiplier: Optional[UnitMultiplier] = Field(default=None, frozen=True)
    unit: Optional[Union[UnitSymbol, Currency]] = Field(default=None, frozen=True)
    denominatorMultiplier: Optional[UnitMultiplier] = Field(default=None, frozen=True)
    denominatorUnit: Optional[UnitSymbol] = Field(default=None, frozen=True)
```